pkg/apis/resource/types.go

```diff
@@ -0,0 +1,388 @@
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. If this is not specified, the resource is
+	// not yet allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
+	//
+	// There can be at most 32 such reservations. This may get increased in
+	// the future, but not reduced.
+	ReservedFor []ResourceClaimConsumerReference
+
+	// DeallocationRequested indicates that a ResourceClaim is to be
+	// deallocated.
+	//
+	// The driver then must deallocate this claim and reset the field
+	// together with clearing the Allocation field.
+	//
+	// While DeallocationRequested is set, no new consumers may be added to
+	// ReservedFor.
+	DeallocationRequested bool
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// claim.status.reservedFor.
+const ResourceClaimReservedForMaxSize = 32
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// ResourceHandle contains arbitrary data returned by the driver after a
+	// successful allocation. This is opaque for
+	// Kubernetes. Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The maximum size of this field is 16KiB. This may get
+	// increased in the future, but not reduced.
+	ResourceHandle string
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// Setting this field is optional. If null, the resource is available
```

Comment:
After the driver is done with the allocation, it knows where the resource is available. For a resource that is reachable via the network, that can be on all nodes. Providing no node selector indicates that.

We cannot use an empty core.NodeSelector for that because that isn't valid content for it (at least one term is required).

Reply:
isn't there an optional path where the driver knows the resource is available on a set of machines (or can learn that) but doesn't have a set of labels to represent the set?

Reply:
Then the node selector must be "node name" in "set of machine names".

Reply:
I'm worried this won't scale, but I am not sure what an answer is other than "set the labels"

Reply:
I'm kind of hoping to see a ClusterResourceSlice API or something like that. Similar to EndpointSlices, a Node could appear in more than one ClusterResourceSlice.

If we have a ClusterResourceSlice API, it _could_ define a `nodeSelector` but perhaps would not have to. A similar design to this, basically. A sliced approach feels like it would scale well at the level of what gets persisted and the volume of _writes_, although it might lead to the scheduler starting more watches.

Reply:
I think this nodeSelector would have to match a topology domain, to say:
this resource can be used in a particular node, rack, zone, etc.

So yes, I would say the answer is to add the labels.